pkg/scheduler/scheduler.go

```diff
@@ -674,8 +683,12 @@ func (sched *Scheduler) scheduleOne(ctx context.Context) {
 		return
 	}
 
+	inSchedulingCycle = false
 	// bind the pod to its host asynchronously (we can do this b/c of the assumption step above).
 	go func() {
+		if sched.SchedulingQueue != nil {
+			defer sched.SchedulingQueue.ClearProcessingPod(pod.UID)
```

Comment:
If we just wait for the update (that includes the node name) from apiserver, we don't even need the defer, right?

But also, if we clear the processing pod in `AddToUnschedulable`, then the only case not covered is successful scheduling, which is only one path. So we could put this call there, instead of using defer that sometimes "gets cancelled".

Reply:
> If we just wait for the update (that includes the node name) from apiserver, we don't even need the defer, right?

I'm not following this. I'm trying to remove the processingPod entry, only in one cycle - either scheduling cycle or binding cycle.

Reply:
I mean that there is the alternative of relaying on the Pod Update coming from the apiserver that adds the node name.

But, if you prefer to treat it just in the scheduling cycle, I believe we don't need the defer. All failure paths already call `sched.recordSchedulingFailure`. The only one that doesn't is the successful scheduling path, in which we can call `ClearProcessingPod` directly.

Reply:
So in total, you're more preferred to remove processingPod in the following 3 places:

- failures path: `AddUnschedulableIfNotPresent`
- some corner cases of failures: [here](https://github.com/kubernetes/kubernetes/blob/dac4f810f5806ea67ee8e9be181e5db0f1a93b29/pkg/scheduler/scheduler.go#L572) and [here](https://github.com/kubernetes/kubernetes/blob/dac4f810f5806ea67ee8e9be181e5db0f1a93b29/pkg/scheduler/scheduler.go#L575)
- happy path: at the end of binding cycle

(we don't need to remove processingPod at the end of the scheduling cycle)

Reply:
Yes. Although, I feel like the 2nd corner case should be refactored somehow. But that's a separate discussion.

Reply:
Done. PTAL.